infra/Pos/Slotting/Types.hs

```diff
@@ -22,15 +38,123 @@ data EpochSlottingData = EpochSlottingData
 
 instance NFData EpochSlottingData
 
+-- Helpful type aliases
+type CurrentEpochSlottingData = EpochSlottingData
+type NextEpochSlottingData    = EpochSlottingData
+
 -- | Data necessary for slotting to work which is basically part of GState.
-data SlottingData = SlottingData
-    { sdPenult      :: !EpochSlottingData
-    -- ^ 'EpochSlottingData' for penult (i. e. 'sdPenultEpoch') epoch.
-    , sdLast        :: !EpochSlottingData
-    -- ^ 'EpochSlottingData' for last epoch.
-    -- Last means 'sdPenultEpoch + 1'.
-    , sdPenultEpoch :: !EpochIndex
-    -- ^ Penult epoch for which 'EpochSlottingData' is known.
-    } deriving (Eq, Show, Generic)
+-- Note that it's important to use error rather than default values like 0, because
+-- such cases indicate invariants violation and shouldn't be hidden behind default values.
+newtype SlottingData = SlottingData
+    { getSlottingDataMap :: Map EpochIndex EpochSlottingData
+    -- ^ Map containing the @EpochSlottingData@ for all the (known) @Epoch@
+    } deriving (Eq, Show, Generic, Monoid)
 
 instance NFData SlottingData
+
+-- | Unsafe constructor that can lead to unsafe crash!
+createSlottingDataUnsafe :: Map EpochIndex EpochSlottingData -> SlottingData
+createSlottingDataUnsafe epochSlottingDataMap =
+    if M.size epochSlottingDataMap < 2
+        then criticalError
+        else SlottingData epochSlottingDataMap
+  where
+    criticalError = error "It's impossible to create slotting data without two epochs."
+
+-- | Restricted constructor function for the (initial) creation of @SlottingData@.
+createInitSlottingData
+    :: CurrentEpochSlottingData
+    -> NextEpochSlottingData
+    -> SlottingData
+createInitSlottingData psd esd = SlottingData validInitialSlottingData
+  where
+
+    validInitialSlottingData :: Map EpochIndex EpochSlottingData
+    validInitialSlottingData = M.union currentEpochSlottingData nextEpochSlottingData
+
+    currentEpochSlottingData :: Map EpochIndex CurrentEpochSlottingData
+    currentEpochSlottingData = M.singleton 0 psd
+
+    nextEpochSlottingData :: Map EpochIndex NextEpochSlottingData
+    nextEpochSlottingData = M.singleton 1 esd
+
+-- | Get all epoch index.
+getAllEpochIndices :: SlottingData -> [EpochIndex]
+getAllEpochIndices = M.keys . getSlottingDataMap
+
+-- | Get the next epoch index.
+getNextEpochIndex :: SlottingData -> EpochIndex
+getNextEpochIndex = fst . M.findMax . getSlottingDataMap
+
+-- | Get the next epoch slotting data.
+getNextEpochSlottingData :: SlottingData -> EpochSlottingData
+getNextEpochSlottingData = snd . M.findMax . getSlottingDataMap
+
+-- | Get the current epoch index. Next epoch - 1.
+getCurrentEpochIndex :: SlottingData -> EpochIndex
+getCurrentEpochIndex = decreaseEpochIndex . getNextEpochIndex
```

Comment:
ok, we can use hashmap if we need to have `O(1)` access (actually I am not sure it's really important) but anyway I suggest validating there is the previous epoch in the map when we put new `EpochSlottingData`